File: lib/src/main/java/com/qwwuyu/lib/gson/EmptyStringAsNullJsonDeserializer.java
package com.qwwuyu.lib.gson;

import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonParseException;
import com.google.gson.JsonPrimitive;

import java.lang.reflect.Type;

/**
 * GSON注解 处理这个动态字段无数据的时候 传递空字符串导致解析出错
 */
class EmptyStringAsNullJsonDeserializer<T> implements JsonDeserializer<T> {
    @Override
    public T deserialize(final JsonElement jsonElement, final Type type, final JsonDeserializationContext context) throws JsonParseException {
        if (!GsonHelper.isTypeString(type)) {
            if (jsonElement.isJsonPrimitive()) {
                final JsonPrimitive jsonPrimitive = jsonElement.getAsJsonPrimitive();
                if (jsonPrimitive.isString() && jsonPrimitive.getAsString().isEmpty()) {
                    return null;
                }
            }
        }
        return context.deserialize(jsonElement, type);
    }
}

File: lib/src/main/kotlin/com/qwwuyu/lib/gson/EmptyStringAsNullJsonDeserializer.kt
package com.qwwuyu.lib.gson

import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import com.google.gson.JsonParseException
import com.qwwuyu.lib.gson.GsonHelper
import java.lang.reflect.Type

/**
 * GSON注解 处理这个动态字段无数据的时候 传递空字符串导致解析出错
 */
internal class EmptyStringAsNullJsonDeserializer<T> : JsonDeserializer<T?> {
    @Throws(JsonParseException::class)
    override fun deserialize(jsonElement: JsonElement, type: Type, context: JsonDeserializationContext): T? {
        if (!GsonHelper.isTypeString(type)) {
            if (jsonElement.isJsonPrimitive) {
                val jsonPrimitive = jsonElement.asJsonPrimitive
                if (jsonPrimitive.isString && jsonPrimitive.asString.isEmpty()) {
                    return null
                }
            }
        }
        return context.deserialize(jsonElement, type)
    }
}